{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "from cinemaot import sinkhorn_knopp as skp\n",
    "#from . import utils\n",
    "from scipy.sparse import issparse\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "import scipy.stats as ss\n",
    "\n",
    "# In this newer version we use the Python implementation of xicor\n",
    "# import rpy2.robjects as ro\n",
    "# import rpy2.robjects.numpy2ri\n",
    "# import rpy2.robjects.pandas2ri\n",
    "# from rpy2.robjects.packages import importr\n",
    "# rpy2.robjects.numpy2ri.activate()\n",
    "# rpy2.robjects.pandas2ri.activate()\n",
    "\n",
    "\n",
    "# Instead of projecting the whole count matrix, we use the pca result of projected ICA components to stablize the noise\n",
    "# returning an anndata object\n",
    "# Detecting differently expressed genes: G = A + Z + AZ + e by NB regression. Significant coefficient before AZ means conditional-specific effect\n",
    "# Further exclusion of false positives may be removed by permutation (as in PseudotimeDE)\n",
    "\n",
    "#import ot\n",
    "\n",
    "import statsmodels.api as sm\n",
    "from sklearn.linear_model import LinearRegression\n",
    "\n",
    "from sklearn.decomposition import FastICA\n",
    "import sklearn.metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def weighted_quantile(values, num, sample_weight=None, \n",
    "                      values_sorted=False):\n",
    "    \"\"\"\n",
    "    Estimate weighted quantile for robust estimation of gene expression change given the OT map. The function is completely vectorized to accelerate computation\n",
    "    \"\"\"\n",
    "    values = np.array(values)\n",
    "    if sample_weight is None:\n",
    "        sample_weight = np.ones(len(values))\n",
    "    sorter = np.argsort(values,axis=0)\n",
    "    values = np.take_along_axis(values, sorter, axis=0)\n",
    "    sample_weight = np.tile(sample_weight/np.sum(sample_weight),(1,values.shape[1]))\n",
    "    sample_weight = np.take_along_axis(sample_weight,sorter,axis=1)\n",
    "    weighted_quantiles = np.cumsum(sample_weight,axis=0)\n",
    "    weighted_quantiles = np.vstack((np.zeros(values.shape[1]),weighted_quantiles))\n",
    "    numindex = np.sum(values <= num.reshape(1,-1),axis=0)\n",
    "    return np.diag(weighted_quantiles[np.ix_(numindex,np.arange(values.shape[1]))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Xi:\n",
    "    \"\"\"\n",
    "    x and y are the data vectors\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, x, y):\n",
    "\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "\n",
    "    @property\n",
    "    def sample_size(self):\n",
    "        return len(self.x)\n",
    "\n",
    "    @property\n",
    "    def x_ordered_rank(self):\n",
    "        # PI is the rank vector for x, with ties broken at random\n",
    "        # Not mine: source (https://stackoverflow.com/a/47430384/1628971)\n",
    "        # random shuffling of the data - reason to use random.choice is that\n",
    "        # pd.sample(frac=1) uses the same randomizing algorithm\n",
    "        len_x = len(self.x)\n",
    "        randomized_indices = np.random.choice(np.arange(len_x), len_x, replace=False)\n",
    "        randomized = [self.x[idx] for idx in randomized_indices]\n",
    "        # same as pandas rank method 'first'\n",
    "        rankdata = ss.rankdata(randomized, method=\"ordinal\")\n",
    "        # Reindexing based on pairs of indices before and after\n",
    "        unrandomized = [\n",
    "            rankdata[j] for i, j in sorted(zip(randomized_indices, range(len_x)))\n",
    "        ]\n",
    "        return unrandomized\n",
    "\n",
    "    @property\n",
    "    def y_rank_max(self):\n",
    "        # f[i] is number of j s.t. y[j] <= y[i], divided by n.\n",
    "        return ss.rankdata(self.y, method=\"max\") / self.sample_size\n",
    "\n",
    "    @property\n",
    "    def g(self):\n",
    "        # g[i] is number of j s.t. y[j] >= y[i], divided by n.\n",
    "        return ss.rankdata([-i for i in self.y], method=\"max\") / self.sample_size\n",
    "\n",
    "    @property\n",
    "    def x_ordered(self):\n",
    "        # order of the x's, ties broken at random.\n",
    "        return np.argsort(self.x_ordered_rank)\n",
    "\n",
    "    @property\n",
    "    def x_rank_max_ordered(self):\n",
    "        x_ordered_result = self.x_ordered\n",
    "        y_rank_max_result = self.y_rank_max\n",
    "        # Rearrange f according to ord.\n",
    "        return [y_rank_max_result[i] for i in x_ordered_result]\n",
    "\n",
    "    @property\n",
    "    def mean_absolute(self):\n",
    "        x1 = self.x_rank_max_ordered[0 : (self.sample_size - 1)]\n",
    "        x2 = self.x_rank_max_ordered[1 : self.sample_size]\n",
    "        \n",
    "        return (\n",
    "            np.mean(\n",
    "                np.abs(\n",
    "                    [\n",
    "                        x - y\n",
    "                        for x, y in zip(\n",
    "                            x1,\n",
    "                            x2,\n",
    "                        )\n",
    "                    ]\n",
    "                )\n",
    "            )\n",
    "            * (self.sample_size - 1)\n",
    "            / (2 * self.sample_size)\n",
    "        )\n",
    "\n",
    "    @property\n",
    "    def inverse_g_mean(self):\n",
    "        gvalue = self.g\n",
    "        return np.mean(gvalue * (1 - gvalue))\n",
    "\n",
    "    @property\n",
    "    def correlation(self):\n",
    "        \"\"\"xi correlation\"\"\"\n",
    "        return 1 - self.mean_absolute / self.inverse_g_mean\n",
    "\n",
    "    @classmethod\n",
    "    def xi(cls, x, y):\n",
    "        return cls(x, y)\n",
    "\n",
    "    def pval_asymptotic(self, ties=False, nperm=1000):\n",
    "        \"\"\"\n",
    "        Returns p values of the correlation\n",
    "        Args:\n",
    "            ties: boolean\n",
    "                If ties is true, the algorithm assumes that the data has ties\n",
    "                and employs the more elaborated theory for calculating\n",
    "                the P-value. Otherwise, it uses the simpler theory. There is\n",
    "                no harm in setting tiles True, even if there are no ties.\n",
    "            nperm: int\n",
    "                The number of permutations for the permutation test, if needed.\n",
    "                default 1000\n",
    "        Returns:\n",
    "            p value\n",
    "        \"\"\"\n",
    "        # If there are no ties, return xi and theoretical P-value:\n",
    "\n",
    "        if ties:\n",
    "            return 1 - ss.norm.cdf(\n",
    "                np.sqrt(self.sample_size) * self.correlation / np.sqrt(2 / 5)\n",
    "            )\n",
    "\n",
    "        # If there are ties, and the theoretical method\n",
    "        # is to be used for calculation P-values:\n",
    "        # The following steps calculate the theoretical variance\n",
    "        # in the presence of ties:\n",
    "        sorted_ordered_x_rank = sorted(self.x_rank_max_ordered)\n",
    "\n",
    "        ind = [i + 1 for i in range(self.sample_size)]\n",
    "        ind2 = [2 * self.sample_size - 2 * ind[i - 1] + 1 for i in ind]\n",
    "\n",
    "        a = (\n",
    "            np.mean([i * j * j for i, j in zip(ind2, sorted_ordered_x_rank)])\n",
    "            / self.sample_size\n",
    "        )\n",
    "\n",
    "        c = (\n",
    "            np.mean([i * j for i, j in zip(ind2, sorted_ordered_x_rank)])\n",
    "            / self.sample_size\n",
    "        )\n",
    "\n",
    "        cq = np.cumsum(sorted_ordered_x_rank)\n",
    "\n",
    "        m = [\n",
    "            (i + (self.sample_size - j) * k) / self.sample_size\n",
    "            for i, j, k in zip(cq, ind, sorted_ordered_x_rank)\n",
    "        ]\n",
    "\n",
    "        b = np.mean([np.square(i) for i in m])\n",
    "        v = (a - 2 * b + np.square(c)) / np.square(self.inverse_g_mean)\n",
    "\n",
    "        return 1 - ss.norm.cdf(\n",
    "            np.sqrt(self.sample_size) * self.correlation / np.sqrt(v)\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cinemaot_unweighted(adata,obs_label,ref_label,expr_label,dim=20,thres=0.15,smoothness=1e-4,eps=1e-3,mode='parametric',marker=None,preweight_label=None):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata: 'AnnData'\n",
    "        An anndata object containing the whole gene count matrix and an observation index for treatments. It should be preprocessed before input.\n",
    "    obs_label: 'str'\n",
    "        A string for indicating the treatment column name in adata.obs.\n",
    "    ref_label: 'str'\n",
    "        A string for indicating the control group in adata.obs.values.\n",
    "    expr_label: 'str'\n",
    "        A string for indicating the experiment group in adata.obs.values.\n",
    "    dim: 'int'\n",
    "        The number of independent components.\n",
    "    thres: 'float'\n",
    "        The threshold for setting the Chatterjee coefficent for confounder separation.\n",
    "    smoothness: 'float'\n",
    "        The parameter for setting the smoothness of entropy-regularized optimal transport. Should be set as a small value above zero!\n",
    "    eps: 'float'\n",
    "        The parameter for stop condition of OT convergence. \n",
    "    mode: 'str'\n",
    "        If mode is 'parametric', return standard differential matrices. If it's non-parametric, we return expr cells' weighted quantile.\n",
    "    Return\n",
    "    ----------\n",
    "    cf: 'numpy.ndarray'\n",
    "        Confounder components, of shape (n_cells,n_components).\n",
    "    ot: 'numpy.ndarray'\n",
    "        Transport map across control and experimental conditions.\n",
    "    te2: 'numpy.ndarray'\n",
    "        Single-cell differential expression for each cell in control condition, of shape (n_refcells, n_genes).\n",
    "    \"\"\"\n",
    "    if dim is None:\n",
    "        sk = skp.SinkhornKnopp()\n",
    "        c = 0.5\n",
    "        data=adata.X\n",
    "        vm = (1e-3 + data + c * data * data)/(1+c)\n",
    "        P = sk.fit(vm)\n",
    "        wm = np.dot(np.dot(np.sqrt(sk._D1),vm),np.sqrt(sk._D2))\n",
    "        u,s,vt = np.linalg.svd(wm)\n",
    "        dim = np.min(sum(s > (np.sqrt(data.shape[0])+np.sqrt(data.shape[1]))),adata.obsm['X_pca'].shape[1])\n",
    "\n",
    "\n",
    "    transformer = FastICA(n_components=dim, random_state=0,whiten=\"arbitrary-variance\")\n",
    "    X_transformed = transformer.fit_transform(adata.obsm['X_pca'][:,:dim])\n",
    "    #importr(\"XICOR\")\n",
    "    #xicor = ro.r[\"xicor\"]\n",
    "    groupvec = (adata.obs[obs_label]==ref_label *1).values #control\n",
    "    xi = np.zeros(dim)\n",
    "    #pval = np.zeros(dim)\n",
    "    j = 0\n",
    "    for source_row in X_transformed.T:\n",
    "        xi_obj = Xi(source_row,groupvec*1)\n",
    "        #rresults = xicor(ro.FloatVector(source_row), ro.FloatVector(groupvec), pvalue = True)\n",
    "        #xi[j] = np.array(rresults.rx2(\"xi\"))[0]\n",
    "        xi[j] = xi_obj.correlation\n",
    "        #pval[j] = np.array(rresults.rx2(\"pval\"))[0]\n",
    "        j = j+1\n",
    "    cf = X_transformed[:,xi<thres]\n",
    "    cf1 = cf[adata.obs[obs_label]==expr_label,:] #expr\n",
    "    cf2 = cf[adata.obs[obs_label]==ref_label,:] #control\n",
    "    if sum(xi<thres)==1:\n",
    "        dis = sklearn.metrics.pairwise_distances(cf1.reshape(-1,1),cf2.reshape(-1,1))\n",
    "    elif sum(xi<thres)==0:\n",
    "        raise ValueError(\"No confounder components identified. Please try a higher threshold.\")\n",
    "    else:\n",
    "        dis = sklearn.metrics.pairwise_distances(cf1,cf2)\n",
    "    e = smoothness * sum(xi<thres)\n",
    "    af = np.exp(-dis * dis / e)\n",
    "    r = np.zeros([cf1.shape[0],1])\n",
    "    c = np.zeros([cf2.shape[0],1])\n",
    "    if preweight_label is None:\n",
    "        r[:,0] = 1/cf1.shape[0]\n",
    "        c[:,0] = 1/cf2.shape[0]\n",
    "    else:\n",
    "        #implement a simple function here, taking adata.obs, output inverse prob weight. For consistency, c is still the empirical distribution, while r is weighted.\n",
    "        adata1 = adata[adata.obs[obs_label]==expr_label,:]\n",
    "        adata2 = adata[adata.obs[obs_label]==ref_label,:]\n",
    "        c[:,0] = 1/cf2.shape[0]\n",
    "        for ct in list(set(adata1.obs[preweight_label].values.tolist())):\n",
    "            r[(adata1.obs[preweight_label]==ct).values,0] = np.sum((adata2.obs[preweight_label]==ct).values) / np.sum((adata1.obs[preweight_label]==ct).values)\n",
    "        r[:,0] = r[:,0]/np.sum(r[:,0])\n",
    "\n",
    "    sk = skp.SinkhornKnopp(setr=r,setc=c,epsilon=eps)\n",
    "    ot_matrix = sk.fit(af).T\n",
    "\n",
    "    embedding = X_transformed[adata.obs[obs_label]==ref_label,:] - np.matmul(ot_matrix/np.sum(ot_matrix,axis=1)[:,None],X_transformed[adata.obs[obs_label]==expr_label,:])\n",
    "\n",
    "    if mode == 'parametric':\n",
    "        if issparse(adata.X):\n",
    "            te2 = adata.X.toarray()[adata.obs[obs_label]==ref_label,:] - np.matmul(ot_matrix/np.sum(ot_matrix,axis=1)[:,None],adata.X.toarray()[adata.obs[obs_label]==expr_label,:])\n",
    "        else:\n",
    "            te2 = adata.X[adata.obs[obs_label]==ref_label,:] - np.matmul(ot_matrix/np.sum(ot_matrix,axis=1)[:,None],adata.X[adata.obs[obs_label]==expr_label,:])\n",
    "    elif mode == 'non_parametric':\n",
    "        if issparse(adata.X):\n",
    "            ref = adata.X.toarray()[adata.obs[obs_label]==ref_label,:]\n",
    "            ref = ref[:,adata.var_names.isin(marker)]\n",
    "            expr = adata.X.toarray()[adata.obs[obs_label]==expr_label,:]\n",
    "            expr = expr[:,adata.var_names.isin(marker)]\n",
    "            te2 = ref * 0\n",
    "            for i in range(te2.shape[0]):\n",
    "                te2[i,:] = weighted_quantile(expr,ref[i,:],sample_weight=ot_matrix[i,:])\n",
    "        else:\n",
    "            ref = adata.X[adata.obs[obs_label]==ref_label,:]\n",
    "            ref = ref[:,adata.var_names.isin(marker)]\n",
    "            expr = adata.X[adata.obs[obs_label]==expr_label,:]\n",
    "            expr = expr[:,adata.var_names.isin(marker)]\n",
    "            te2 = ref * 0\n",
    "            for i in range(te2.shape[0]):\n",
    "                te2[i,:] = weighted_quantile(expr,ref[i,:],sample_weight=ot_matrix[i,:])            \n",
    "    else:\n",
    "        raise ValueError(\"We do not support other methods for DE now.\")\n",
    "\n",
    "    TE = sc.AnnData(te2,obs=adata[adata.obs[obs_label]==ref_label,:].obs.copy(),var=adata.var.copy())\n",
    "    TE.obsm['X_embedding'] = embedding\n",
    "    return cf, ot_matrix, TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('data/Integrated_subset.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arturdox/.virtualenvs/CINEMA-OT/lib/python3.12/site-packages/scanpy/preprocessing/_pca.py:229: ImplicitModificationWarning: Setting element `.obsm['X_pca']` of view, initializing view as actual.\n",
      "  adata.obsm['X_pca'] = X_pca\n",
      "/Users/arturdox/.virtualenvs/CINEMA-OT/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:128: ConvergenceWarning: FastICA did not converge. Consider increasing tolerance or the maximum number of iterations.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "adata_ = adata[adata.obs[\"perturbation\"].isin([\"No stimulation\", \"IFNb\"])]\n",
    "sc.pp.pca(adata_)\n",
    "cf, ot, de = cinemaot_unweighted(\n",
    "    adata_,\n",
    "    obs_label=\"perturbation\",\n",
    "    ref_label=\"No stimulation\",\n",
    "    expr_label=\"IFNb\",\n",
    "    mode=\"parametric\",\n",
    "    thres=0.5,\n",
    "    smoothness=1e-5,\n",
    "    eps=1e-3,\n",
    "    preweight_label=\"cell_type0528\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.04288645e-17, 4.39323647e-14, 8.65123580e-15, ...,\n",
       "        2.71220876e-15, 8.25183199e-19, 1.14311443e-12],\n",
       "       [1.19102212e-32, 4.26832062e-22, 1.02291188e-23, ...,\n",
       "        3.00446440e-23, 2.77416264e-22, 5.05143041e-28],\n",
       "       [3.02779461e-11, 2.92344022e-08, 2.91644610e-08, ...,\n",
       "        6.01648001e-15, 4.85040164e-08, 4.25144831e-11],\n",
       "       ...,\n",
       "       [2.11124543e-23, 2.35593573e-22, 6.76845508e-23, ...,\n",
       "        5.14545000e-26, 8.24019216e-21, 5.89943866e-16],\n",
       "       [1.40929510e-12, 1.24397978e-14, 1.86974088e-09, ...,\n",
       "        9.03585371e-18, 3.75657794e-14, 1.29347429e-17],\n",
       "       [1.47324371e-15, 4.47141112e-11, 4.68498025e-17, ...,\n",
       "        3.43331379e-15, 8.08898688e-17, 1.00301695e-16]])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>perturbation</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>batch</th>\n",
       "      <th>leiden</th>\n",
       "      <th>cell_type0528</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCTAGCA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>938</td>\n",
       "      <td>938</td>\n",
       "      <td>1921.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>4.060385</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTGGAGTC-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>4288</td>\n",
       "      <td>4287</td>\n",
       "      <td>16910.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>1.632170</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>18</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACATCTTT-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>691</td>\n",
       "      <td>691</td>\n",
       "      <td>1199.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>3.336113</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>14</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATTATCTC-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>1072</td>\n",
       "      <td>1072</td>\n",
       "      <td>2590.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>2.972973</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGAGACTAGGC-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>781</td>\n",
       "      <td>781</td>\n",
       "      <td>1345.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>2.007435</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACAACTGT-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>863</td>\n",
       "      <td>863</td>\n",
       "      <td>1654.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>2.781137</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGAAATA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>3239</td>\n",
       "      <td>3238</td>\n",
       "      <td>10030.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>2.582253</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>12</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTCTGATCA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>1426</td>\n",
       "      <td>1425</td>\n",
       "      <td>2727.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>2.566923</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>22</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCTATTCA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>486</td>\n",
       "      <td>486</td>\n",
       "      <td>646.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>4.024768</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>23</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTGTACT-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>1034</td>\n",
       "      <td>1033</td>\n",
       "      <td>2126.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>2.916275</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2268 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      perturbation  n_genes  n_genes_by_counts  total_counts  \\\n",
       "AAACCTGAGGCTAGCA-1  No stimulation      938                938        1921.0   \n",
       "AAACCTGAGTGGAGTC-1  No stimulation     4288               4287       16910.0   \n",
       "AAACCTGCACATCTTT-1  No stimulation      691                691        1199.0   \n",
       "AAACCTGCATTATCTC-1  No stimulation     1072               1072        2590.0   \n",
       "AAACGGGAGACTAGGC-1  No stimulation      781                781        1345.0   \n",
       "...                            ...      ...                ...           ...   \n",
       "TTTGTCACACAACTGT-1  No stimulation      863                863        1654.0   \n",
       "TTTGTCACACGAAATA-1  No stimulation     3239               3238       10030.0   \n",
       "TTTGTCAGTCTGATCA-1  No stimulation     1426               1425        2727.0   \n",
       "TTTGTCATCCTATTCA-1  No stimulation      486                486         646.0   \n",
       "TTTGTCATCTTGTACT-1  No stimulation     1034               1033        2126.0   \n",
       "\n",
       "                    total_counts_mt  pct_counts_mt batch leiden cell_type0528  \n",
       "AAACCTGAGGCTAGCA-1             78.0       4.060385  H3D2      7         CD8 T  \n",
       "AAACCTGAGTGGAGTC-1            276.0       1.632170  H3D2     18      Monocyte  \n",
       "AAACCTGCACATCTTT-1             40.0       3.336113  H3D2     14         CD4 T  \n",
       "AAACCTGCATTATCTC-1             77.0       2.972973  H3D2      0         CD4 T  \n",
       "AAACGGGAGACTAGGC-1             27.0       2.007435  H3D2      2         CD8 T  \n",
       "...                             ...            ...   ...    ...           ...  \n",
       "TTTGTCACACAACTGT-1             46.0       2.781137  H3D2      7         CD8 T  \n",
       "TTTGTCACACGAAATA-1            259.0       2.582253  H3D2     12      Monocyte  \n",
       "TTTGTCAGTCTGATCA-1             70.0       2.566923  H3D2     22            NK  \n",
       "TTTGTCATCCTATTCA-1             26.0       4.024768  H3D2     23         CD4 T  \n",
       "TTTGTCATCTTGTACT-1             62.0       2.916275  H3D2      7         CD8 T  \n",
       "\n",
       "[2268 rows x 9 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_.obs[adata_.obs['perturbation']=='No stimulation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>perturbation</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>batch</th>\n",
       "      <th>leiden</th>\n",
       "      <th>cell_type0528</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATTGTG-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>2124</td>\n",
       "      <td>2124</td>\n",
       "      <td>5351.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>1.233414</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>10</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGATGCCAG-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>1650</td>\n",
       "      <td>1650</td>\n",
       "      <td>4109.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1.727914</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTGCCAG-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>720</td>\n",
       "      <td>720</td>\n",
       "      <td>1104.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.358696</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>5</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTACCGGCT-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>897</td>\n",
       "      <td>897</td>\n",
       "      <td>1746.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>2.233677</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>3</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTCAGAAGC-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>1203</td>\n",
       "      <td>1203</td>\n",
       "      <td>2351.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.361123</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>3</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCGTGGACC-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>797</td>\n",
       "      <td>797</td>\n",
       "      <td>1322.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.983359</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>20</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTTGTTTG-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>864</td>\n",
       "      <td>864</td>\n",
       "      <td>1473.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.900882</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>1</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGCGCTTAT-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>2964</td>\n",
       "      <td>2964</td>\n",
       "      <td>7918.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>1.490275</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>10</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTAAGAGGA-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>1237</td>\n",
       "      <td>1237</td>\n",
       "      <td>2447.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>3.146710</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>3</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAAACCAC-1</th>\n",
       "      <td>IFNb</td>\n",
       "      <td>623</td>\n",
       "      <td>623</td>\n",
       "      <td>971.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.544799</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>5</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2759 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   perturbation  n_genes  n_genes_by_counts  total_counts  \\\n",
       "AAACCTGAGAATTGTG-1         IFNb     2124               2124        5351.0   \n",
       "AAACCTGAGATGCCAG-1         IFNb     1650               1650        4109.0   \n",
       "AAACCTGCACTGCCAG-1         IFNb      720                720        1104.0   \n",
       "AAACCTGGTACCGGCT-1         IFNb      897                897        1746.0   \n",
       "AAACCTGGTCAGAAGC-1         IFNb     1203               1203        2351.0   \n",
       "...                         ...      ...                ...           ...   \n",
       "TTTGGTTTCGTGGACC-1         IFNb      797                797        1322.0   \n",
       "TTTGGTTTCTTGTTTG-1         IFNb      864                864        1473.0   \n",
       "TTTGTCAAGCGCTTAT-1         IFNb     2964               2964        7918.0   \n",
       "TTTGTCAGTAAGAGGA-1         IFNb     1237               1237        2447.0   \n",
       "TTTGTCATCAAACCAC-1         IFNb      623                623         971.0   \n",
       "\n",
       "                    total_counts_mt  pct_counts_mt batch leiden cell_type0528  \n",
       "AAACCTGAGAATTGTG-1             66.0       1.233414  H3D2     10      Monocyte  \n",
       "AAACCTGAGATGCCAG-1             71.0       1.727914  H3D2      0         CD4 T  \n",
       "AAACCTGCACTGCCAG-1             15.0       1.358696  H3D2      5            NK  \n",
       "AAACCTGGTACCGGCT-1             39.0       2.233677  H3D2      3         CD4 T  \n",
       "AAACCTGGTCAGAAGC-1             32.0       1.361123  H3D2      3         CD4 T  \n",
       "...                             ...            ...   ...    ...           ...  \n",
       "TTTGGTTTCGTGGACC-1             13.0       0.983359  H3D2     20         CD4 T  \n",
       "TTTGGTTTCTTGTTTG-1             28.0       1.900882  H3D2      1         CD4 T  \n",
       "TTTGTCAAGCGCTTAT-1            118.0       1.490275  H3D2     10      Monocyte  \n",
       "TTTGTCAGTAAGAGGA-1             77.0       3.146710  H3D2      3         CD4 T  \n",
       "TTTGTCATCAAACCAC-1             15.0       1.544799  H3D2      5            NK  \n",
       "\n",
       "[2759 rows x 9 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_.obs[adata_.obs['perturbation']=='IFNb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>perturbation</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>batch</th>\n",
       "      <th>leiden</th>\n",
       "      <th>cell_type0528</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCTAGCA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>938</td>\n",
       "      <td>938</td>\n",
       "      <td>1921.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>4.060385</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTGGAGTC-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>4288</td>\n",
       "      <td>4287</td>\n",
       "      <td>16910.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>1.632170</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>18</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACATCTTT-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>691</td>\n",
       "      <td>691</td>\n",
       "      <td>1199.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>3.336113</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>14</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATTATCTC-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>1072</td>\n",
       "      <td>1072</td>\n",
       "      <td>2590.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>2.972973</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGAGACTAGGC-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>781</td>\n",
       "      <td>781</td>\n",
       "      <td>1345.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>2.007435</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>2</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACAACTGT-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>863</td>\n",
       "      <td>863</td>\n",
       "      <td>1654.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>2.781137</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGAAATA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>3239</td>\n",
       "      <td>3238</td>\n",
       "      <td>10030.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>2.582253</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>12</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTCTGATCA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>1426</td>\n",
       "      <td>1425</td>\n",
       "      <td>2727.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>2.566923</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>22</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCTATTCA-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>486</td>\n",
       "      <td>486</td>\n",
       "      <td>646.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>4.024768</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>23</td>\n",
       "      <td>CD4 T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTGTACT-1</th>\n",
       "      <td>No stimulation</td>\n",
       "      <td>1034</td>\n",
       "      <td>1033</td>\n",
       "      <td>2126.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>2.916275</td>\n",
       "      <td>H3D2</td>\n",
       "      <td>7</td>\n",
       "      <td>CD8 T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2268 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      perturbation  n_genes  n_genes_by_counts  total_counts  \\\n",
       "AAACCTGAGGCTAGCA-1  No stimulation      938                938        1921.0   \n",
       "AAACCTGAGTGGAGTC-1  No stimulation     4288               4287       16910.0   \n",
       "AAACCTGCACATCTTT-1  No stimulation      691                691        1199.0   \n",
       "AAACCTGCATTATCTC-1  No stimulation     1072               1072        2590.0   \n",
       "AAACGGGAGACTAGGC-1  No stimulation      781                781        1345.0   \n",
       "...                            ...      ...                ...           ...   \n",
       "TTTGTCACACAACTGT-1  No stimulation      863                863        1654.0   \n",
       "TTTGTCACACGAAATA-1  No stimulation     3239               3238       10030.0   \n",
       "TTTGTCAGTCTGATCA-1  No stimulation     1426               1425        2727.0   \n",
       "TTTGTCATCCTATTCA-1  No stimulation      486                486         646.0   \n",
       "TTTGTCATCTTGTACT-1  No stimulation     1034               1033        2126.0   \n",
       "\n",
       "                    total_counts_mt  pct_counts_mt batch leiden cell_type0528  \n",
       "AAACCTGAGGCTAGCA-1             78.0       4.060385  H3D2      7         CD8 T  \n",
       "AAACCTGAGTGGAGTC-1            276.0       1.632170  H3D2     18      Monocyte  \n",
       "AAACCTGCACATCTTT-1             40.0       3.336113  H3D2     14         CD4 T  \n",
       "AAACCTGCATTATCTC-1             77.0       2.972973  H3D2      0         CD4 T  \n",
       "AAACGGGAGACTAGGC-1             27.0       2.007435  H3D2      2         CD8 T  \n",
       "...                             ...            ...   ...    ...           ...  \n",
       "TTTGTCACACAACTGT-1             46.0       2.781137  H3D2      7         CD8 T  \n",
       "TTTGTCACACGAAATA-1            259.0       2.582253  H3D2     12      Monocyte  \n",
       "TTTGTCAGTCTGATCA-1             70.0       2.566923  H3D2     22            NK  \n",
       "TTTGTCATCCTATTCA-1             26.0       4.024768  H3D2     23         CD4 T  \n",
       "TTTGTCATCTTGTACT-1             62.0       2.916275  H3D2      7         CD8 T  \n",
       "\n",
       "[2268 rows x 9 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "control = adata[adata.obs['perturbation']=='No stimulation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(ArrayView([[-1.5649903e-01,  4.6627760e+00, -1.3260117e-01, ...,\n",
       "             -1.9857515e-01, -1.6133374e-01, -2.3858619e-01],\n",
       "            [ 2.5749395e-02,  6.0544107e-02,  7.2225355e-02, ...,\n",
       "             -2.3227680e-02,  9.0107147e-04,  2.1258122e-01],\n",
       "            [-1.6670781e-01, -2.8013307e-01, -1.5954842e-01, ...,\n",
       "             -2.1336092e-01, -1.6441849e-01, -2.6941788e-01],\n",
       "            ...,\n",
       "            [-1.4901598e-01, -2.4170154e-01, -1.4926013e-01, ...,\n",
       "             -1.9941691e-01, -1.4494702e-01,  4.4739728e+00],\n",
       "            [-1.7241320e-01, -2.9764402e-01, -1.5494134e-01, ...,\n",
       "             -2.1531560e-01, -1.7377223e-01, -2.7958304e-01],\n",
       "            [-1.5589504e-01, -2.5722858e-01, -1.5235630e-01, ...,\n",
       "             -2.0454866e-01, -1.5286881e-01, -2.4443221e-01]],\n",
       "           dtype=float32),\n",
       " (2268, 773))"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control.X, control.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[-0.11865562, -0.17559336, -0.13184807, ..., -0.17556609,\n",
       "         -0.11143828, -0.1599121 ],\n",
       "        [-0.13325824, -0.20568384, -0.14286487, ..., -0.18788525,\n",
       "         -0.12653024, -0.19412361],\n",
       "        [-0.15649903,  4.662776  , -0.13260117, ..., -0.19857515,\n",
       "         -0.16133374, -0.23858619],\n",
       "        ...,\n",
       "        [-0.12380419, -0.19445087, -0.12295869, ..., -0.17581214,\n",
       "         -0.12171488, -0.1673852 ],\n",
       "        [-0.1724132 , -0.29764402, -0.15494134, ..., -0.2153156 ,\n",
       "         -0.17377223, -0.27958304],\n",
       "        [-0.15589504, -0.25722858, -0.1523563 , ..., -0.20454866,\n",
       "         -0.1528688 , -0.24443221]], dtype=float32),\n",
       " (9209, 773))"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X, adata.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9209 × 773\n",
       "    obs: 'perturbation', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'batch', 'leiden', 'cell_type0528'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'batch_colors', 'cell_type0528_colors', 'hvg', 'leiden', 'leiden_colors', 'log1p', 'neighbors', 'pca', 'perturbation_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[-5.36350218e-02,  4.79445373e+00, -2.50692672e+00, ...,\n",
       "         -1.75557896e-04, -8.15729347e-03,  3.74461862e-02],\n",
       "        [ 2.13709679e-02, -1.05344196e-02, -4.70492095e-02, ...,\n",
       "         -2.19506462e-01, -3.78152651e-02,  4.16946664e-02],\n",
       "        [-6.19409904e-02, -1.91294310e-01, -8.45040081e-03, ...,\n",
       "         -7.44984073e-02, -8.68032782e-02, -5.29177999e-02],\n",
       "        ...,\n",
       "        [-3.89115187e-01, -1.30004209e+00, -1.24121167e+00, ...,\n",
       "         -2.06752127e-01, -4.69588964e-01,  2.47418985e+00],\n",
       "        [-3.11096077e-03,  6.06325101e-04, -1.24358385e-02, ...,\n",
       "         -3.39462535e-01,  3.99766515e-04, -9.66897429e-03],\n",
       "        [-3.42868360e-02, -2.52850823e-01, -7.84349447e-01, ...,\n",
       "         -2.72645833e-02, -1.42368016e-02,  1.18387235e-02]]),\n",
       " (2268, 773))"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de.X, de.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CINEMA-OT",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
